src/go/cmd/http-relay-server/server/broker_test.go

```diff
@@ -221,3 +255,51 @@ func TestTimeout(t *testing.T) {
 	}()
 	wg.Wait()
 }
+
+func TestReapWhileSendingResponse(t *testing.T) {
+	b := newBroker()
+	b.req["foo"] = make(chan *pb.HttpRequest)
+
+	// create a broker connection between the user client and backend side, but don't
+	// start sending data. "req" is backend side and "resp" is user client side.
+	var req *pb.HttpRequest
+	var reqErr error
+	// var resp <-chan *pb.HttpResponse
+	var respErr error
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		req, reqErr = b.GetRequest(context.Background(), "foo", "/")
+		if reqErr != nil {
+			t.Errorf("GetRequest error:", reqErr)
+		}
+		wg.Done()
+	}()
+	go func() {
+		_, respErr = b.RelayRequest("foo", &pb.HttpRequest{Id: proto.String(idOne), Url: proto.String("http://example.com/foo")})
+		if respErr != nil {
+			t.Errorf("RelayRequest error:", respErr)
+		}
+		wg.Done()
+	}()
+	wg.Wait()
+	if reqErr != nil || respErr != nil {
+		t.Errorf("Error making broker connection")
+	}
+
+	// start sending response to user client, BUT do not start consuming the response.
+	wg.Add(1)
+	go func() {
+		reqErr = b.SendResponse(&pb.HttpResponse{Id: req.Id, Body: []byte(*req.Id), Eof: proto.Bool(false)})
+		if reqErr == nil || reqErr.Error() != "Closed due to inactivity" {
+			t.Errorf("Wrong SendResponse error or no error:", reqErr)
+		}
+		wg.Done()
+	}()
+	// FIXME(koonpeng): we need to wait for the goroutinue to be blocked on writing the response, currently
+	// there is no way to confirm that so we use a sleep.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
I would really like to remove this sleep, it makes the test slow and flakey, but due to the racey nature it's almost impossible to do a sync. I also tried running with `--@io_bazel_rules_go//go/config:race` but it also does not catch the race without this sleep.